src/shogun/io/File.cpp

```diff
@@ -76,111 +76,116 @@ CFile::CFile(const char* fname, char rw, const char* name) : CSGObject()
 		set_variable_name(name);
 }
 
-void CFile::get_vector(bool*& vector, int32_t& len)
-{
-	int32_t* int_vector;
-	get_vector(int_vector, len);
-
-	ASSERT(len>0)
-	vector= SG_MALLOC(bool, len);
-
-	for (int32_t i=0; i<len; i++)
-		vector[i]= (int_vector[i]!=0);
-
-	SG_FREE(int_vector);
-}
-
-void CFile::set_vector(const bool* vector, int32_t len)
-{
-	int32_t* int_vector = SG_MALLOC(int32_t, len);
-	for (int32_t i=0;i<len;i++)
-	{
-		if (vector[i])
-			int_vector[i]=1;
-		else
-			int_vector[i]=0;
-	}
-	set_vector(int_vector,len);
-	SG_FREE(int_vector);
-}
-
-void CFile::get_matrix(bool*& matrix, int32_t& num_feat, int32_t& num_vec)
+CFile::~CFile()
 {
-	uint8_t * byte_matrix;
-	get_matrix(byte_matrix,num_feat,num_vec);
-
-	ASSERT(num_feat > 0 && num_vec > 0)
-	matrix = SG_MALLOC(bool, num_feat*num_vec);
-
-	for(int32_t i = 0;i < num_vec;i++)
-	{
-		for(int32_t j = 0;j < num_feat;j++)
-			matrix[i*num_feat+j] = byte_matrix[i*num_feat+j] != 0 ? 1 : 0;
-	}
-
-	SG_FREE(byte_matrix);
+	close();
 }
 
-void CFile::set_matrix(const bool* matrix, int32_t num_feat, int32_t num_vec)
-{
-	uint8_t * byte_matrix = SG_MALLOC(uint8_t, num_feat*num_vec);
-	for(int32_t i = 0;i < num_vec;i++)
-	{
-		for(int32_t j = 0;j < num_feat;j++)
-			byte_matrix[i*num_feat+j] = matrix[i*num_feat+j] != 0 ? 1 : 0;
-	}
-
-	set_matrix(byte_matrix,num_feat,num_vec);
-
-	SG_FREE(byte_matrix);
+#define VECTOR_GETSET(type)                                 \
+void CFile::set_vector(const type* vector, int32_t len)     \
+{                                                           \
+    int32_t* int_vector = SG_MALLOC(int32_t, len);          \
+    for (int32_t i=0;i<len;i++)                             \
+    {                                                       \
+        if (vector[i])                                      \
+            int_vector[i]=1;                                \
+        else                                                \
+            int_vector[i]=0;                                \
+    }                                                       \
+    set_vector(int_vector,len);                             \
```

Comment:
I again checked my implementation. I think its completely right. Here, set_vector is a virtual function. So, when it's called the 2nd time, it will bypass the base class implementation and call the implementation of the derived class. Thus, there wouldn't be any recursion. 
Moreover, after looking more closely, I discovered these are in fact dummy implementation. In all cases, the derived class implementation will be called, unless base class implementation is called explicitly, which would hardly be the case practically.